pkg/singleflight/singleflight.go

```diff
@@ -0,0 +1,106 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package singleflight
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"golang.org/x/sync/semaphore"
+)
+
+// Singleflighter provides suppression control of duplicate function calls.
+type Singleflighter interface {
```

Comment:
Compare to pkg/semaphore vs sync/semaphore, this is sufficiently different from https://pkg.go.dev/golang.org/x/sync/singleflight. It may be better to have a different name.

Isn't it a keyed mutex?